File: core/backend/src/signals/__tests__/moving-average.test.ts
import { describe, expect, it } from "vitest";
import { movingAverage } from "../moving-average.js";

describe("movingAverage", () => {
  it("returns null when below minSamples", () => {
    const r = movingAverage([100, 101, 102], 20, 20);
    expect(r.value).toBeNull();
    expect(r.samplesUsed).toBe(3);
  });

  it("averages exactly windowSize most-recent entries", () => {
    const closes = Array.from({ length: 25 }, (_, i) => 100 + i); // 100..124
    // recent-first order means the first 20 entries are [100..119]
    const r = movingAverage(closes, 20, 20);
    expect(r.value).toBe("109.5000"); // mean of 100..119
    expect(r.samplesUsed).toBe(20);
  });

  it("handles string close prices (Decimal-serialized)", () => {
    const closes = Array.from({ length: 20 }, () => "500.50");
    const r = movingAverage(closes, 20, 20);
    expect(r.value).toBe("500.5000");
    expect(r.samplesUsed).toBe(20);
  });
});

File: core/mobile/test/engine/moving_average_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nepse_buy/domain/engine/moving_average.dart';

void main() {
  group('movingAverage', () {
    test('returns null when fewer than minSamples', () {
      final r = movingAverage([100, 101, 102], windowSize: 20, minSamples: 20);
      expect(r.value, isNull);
      expect(r.samplesUsed, 3);
    });

    test('computes mean of the most recent windowSize entries', () {
      final closes = List<double>.generate(20, (i) => (100 + i).toDouble());
      final r = movingAverage(closes, windowSize: 20, minSamples: 20);
      // mean(100..119) = 109.5
      expect(r.value, 109.5);
      expect(r.samplesUsed, 20);
    });

    test('truncates to windowSize when more samples present', () {
      final closes = [
        for (var i = 0; i < 30; i++) (200 + i).toDouble(),
      ];
      final r = movingAverage(closes, windowSize: 20, minSamples: 20);
      // First 20 (most-recent-first) = 200..219 → mean 209.5
      expect(r.value, 209.5);
    });

    test('rounds to 4 decimals', () {
      final r = movingAverage([1.11111, 2.22222, 3.33333],
          windowSize: 3, minSamples: 3);
      // mean = 2.22222 — already 4 decimals.
      expect(r.value, 2.2222);
    });
  });
}
